Content.Shared/Automod/AutomodActionGroupPrototype.cs

```diff
@@ -0,0 +1,86 @@
+﻿using Content.Shared.Database;
+using JetBrains.Annotations;
+using Robust.Shared.Player;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Automod;
+
+/// <summary>
+/// An automod filter for running actions when a pattern match is found in user input.
+/// </summary>
+/// <param name="pattern">The text to find in user input.</param>
+/// <param name="filterType">The type of filter to sort this definition into.</param>
+/// <param name="actionGroup">The <seealso cref="AutomodActionGroupPrototype"/> to run when the <paramref name="pattern"/> is matched</param>
+/// <param name="targets">The user input types that this filter applies to.</param>
+/// <param name="name">The user facing name of this filter.</param>
+public sealed class AutomodFilterDef(
+    int? id,
+    string pattern,
+    AutomodFilterType filterType,
+    ProtoId<AutomodActionGroupPrototype> actionGroup,
+    AutomodTarget targets,
+    string name)
+{
+    public readonly int? Id = id;
+    public readonly string Pattern = pattern;
+    public readonly AutomodFilterType FilterType = filterType;
+    public readonly string ActionGroup = actionGroup;
+    public readonly AutomodTarget TargetFlags = targets;
+    public readonly string DisplayName = name;
+
+    public AutomodFilterDef(
+        string pattern,
+        AutomodFilterType filterType,
+        string actionGroup,
+        AutomodTarget targets,
+        string name) : this(null, pattern, filterType, actionGroup, targets, name)
+    {
+    }
+}
+
+/// <summary>
+/// A group of actions to be run when the automod pattern matches the text.
+/// </summary>
+[Prototype("automodActionGroup")]
+public sealed class AutomodActionGroupPrototype : IPrototype
+{
+    [ViewVariables]
+    [IdDataField]
+    public string ID { get; private set; } = default!;
+
+    [DataField]
+    public List<ITextAutomodAction> AutomodActions = new();
+}
+
+[UsedImplicitly(ImplicitUseTargetFlags.WithInheritors)]
+[ImplicitDataDefinitionForInheritors]
+public partial interface ITextAutomodAction
```

Comment:
This heavily OOP design immediately falls apart at the seams, as can be seen from the high amounts of IoC resolves needed. This stuff needs to be handled by managers of some kind.

Reply:
Can you suggest a more specific way to do that? I'm coming up blank.

Reply:
Easiest thing is just to do something similar to event bus where managers subscribe to the relevant event based on the type in question. All these events can just be put in a partial of automod action manager to keep stuff simple.